File: backend/app/Models/Role.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsToMany;

class Role extends Model
{
    protected $fillable = ['name', 'display_name', 'description'];

    // Role <-> User (nhiều-nhiều)
    public function users(): BelongsToMany
    {
        return $this->belongsToMany(User::class);
    }

    // Role <-> Permission (nhiều-nhiều)
    public function permissions(): BelongsToMany
    {
        return $this->belongsToMany(Permission::class, 'role_permission');
    }
}

File: backend/models/role.go
package models

import "time"

type Role struct {
	ID          uint         `gorm:"primaryKey" json:"id"`
	CreatedAt   time.Time    `json:"created_at"`
	UpdatedAt   time.Time    `json:"updated_at"`
	Name        string       `gorm:"uniqueIndex:idx_role_name;not null" json:"name"`
	Permissions []Permission `gorm:"many2many:role_permissions;" json:"permissions,omitempty"`
	Users       []User       `gorm:"many2many:role_users;" json:"users,omitempty"`
}

type Permission struct {
	ID        uint      `gorm:"primaryKey" json:"id"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
	Name      string    `gorm:"uniqueIndex:idx_permission_name;not null" json:"name"`
	Roles     []Role    `gorm:"many2many:role_permissions;" json:"roles,omitempty"`
}
